resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/AtomicRateLimiter.java

```diff
@@ -201,25 +202,26 @@ private State calculateNextState(final long timeoutInNanos, final State activeSt
             nextPermissions = (int) min(nextPermissions + accumulatedPermissions, permissionsPerCycle);
         }
         long nextNanosToWait = nanosToWaitForPermission(
-                cyclePeriodInNanos, permissionsPerCycle, nextPermissions, currentNanos, currentCycle
+                weight, cyclePeriodInNanos, permissionsPerCycle, nextPermissions, currentNanos, currentCycle
         );
-        State nextState = reservePermissions(activeState.config, timeoutInNanos, nextCycle, nextPermissions, nextNanosToWait);
+        State nextState = reservePermissions(activeState.config, weight, timeoutInNanos, nextCycle, nextPermissions, nextNanosToWait);
         return nextState;
     }
 
     /**
      * Calculates time to wait for next permission as
      * [time to the next cycle] + [duration of full cycles until reserved permissions expire]
      *
+     * @param weight               weight of required permissions
      * @param cyclePeriodInNanos   current configuration values
      * @param permissionsPerCycle  current configuration values
      * @param availablePermissions currently available permissions, can be negative if some permissions have been reserved
      * @param currentNanos         current time in nanoseconds
      * @param currentCycle         current {@link AtomicRateLimiter} cycle    @return nanoseconds to wait for the next permission
      */
-    private long nanosToWaitForPermission(final long cyclePeriodInNanos, final int permissionsPerCycle,
+    private long nanosToWaitForPermission(final int weight, final long cyclePeriodInNanos, final int permissionsPerCycle,
                                           final int availablePermissions, final long currentNanos, final long currentCycle) {
-        if (availablePermissions > 0) {
+        if (availablePermissions >= weight) {
             return 0L;
         }
         long nextCycleTimeInNanos = (currentCycle + 1) * cyclePeriodInNanos;
```

Comment:
Unfortunately, you can't just update a happy path for this function. With `weight` being introduced to this function, you will need to wait longer, potentially even more than one cycle to accumulate the required amount of permissions.

Reply:
I don't think that having weights > limitForPeriod is an actual use case that exists in the wild. One could argue if a `aquirePermission` in this situation should wait multiples of limitRefreshPeriod or should it always timeout in this situation. Having a `aquirePermission` that block even on a fresh cycle seams strange to me.

I would suggest to address this issue either by:
a) throwing an exception the passed weight is weights > limitForPeriod
b) adding some documentation that which ever weight you pass we will not wait more then one limitRefreshPeriod

Reply:
so, in which direction do we want to take this?
should we have an `aquirePermission` that can wait even on a fresh cycle or should we do (a) or (b) from my proposals?
personally I'm in favour of (a)

Reply:
I believe that we should not change the contract and implement current semantic "if an `aquirePermission` in this situation should wait for a multiple of `limitRefreshPeriod's`". The client already can pass the expected timeout, so if `nanosToWaitForPermission` returns too big duration, we won't reserve these permissions.

Reply:
so...
even if the rate limiter did do nothing for the past dozens of cycles,
if I call `aquirePermission(6)` and my `limitForPeriod = 2`,
then you want me to wait for 3 cycles?

this behaviour is absurd but if you want it I'll implement it
from my perspective its an error to have weight > limitForPeriod and I would never allow it in my application code nor does any API service known to me impose such bizarre rules - they would be even impossible to enforce from the APIs perspective because you'll wait even before making your first call to a function with a weight > limitForPeriod ...

Reply:
Hi @walec51

First of all, I want to say that I'm very happy and thankful that you decided to contribute such a change to our library. I think that this is a good idea and it will be valuable for some of our users.
But I want to kindly ask you to try to avoid any subjective thoughts and messages. Things like "so..." and "this behavior is absurd" can be considered rude by other contributors.

As for your point about, "its an error to have `weight` > `limitForPeriod` and I would never allow it in my application code". I totally get it, especially if we consider only "blocking" APIs.
But current AtomicRateLimiter is essentially an implementation of a super simple "fixed window rate limiting" algorithm with non-blocking API. And in our implementation, it is OK for `State.activePermissions` to be a negative `int`, where it means that there is a thread or multiple threads that reserved some permissions for future calls and are waiting for their turn or doing some other work while waiting. 
So when we take this into consideration, there can be a situation where
`weight` is less than `limitForPeriod`, but you'll need to wait for more than one cycle for your turn to perform a request.

Hopefully, this makes sense for you, and please correct me if I'm wrong.

Reply:
ok, sorry for my attitude, please take a look at the current implementation